akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -145,29 +144,56 @@ private[fusing] final class LocalSinkSubscription[T](sub: ActorPublisherMessage
 /**
  * INTERNAL API
  */
-private[fusing] final class LocalSource[T](var sub: LocalSinkSubscription[T] = null, var elem: T = null.asInstanceOf[T])
+private[fusing] final class LocalSource[T] {
+  private var subF: Future[LocalSinkSubscription[T]] = _
+  private var sub: LocalSinkSubscription[T] = _
+
+  var elem: T = null.asInstanceOf[T]
+
+  def isActive: Boolean = sub ne null
+  def deactivate(): Unit = sub = null
+  def activate(f: Future[LocalSinkSubscription[T]]): Unit = {
+    subF = f
+    f.foreach(s ⇒ sub = s)(sameThreadExecutionContext)
+  }
+
+  def pull(): Unit = {
+    if (sub eq null)
+      throw new IllegalStateException("not yet initialized, subF=" + subF.value)
+    sub.pull()
+  }
+
+  def cancel(): Unit = subF.foreach(_.cancel())(sameThreadExecutionContext)
```

Comment:
This is why this is an implementation detail of FlattenMerge, which you are about to properly express by moving things into a tighter namespace (thanks!).